### PR TITLE
historian - keep N versions in memory

### PR DESCRIPTION
**Pull request by indera-shsp**
Signed-off-by: Andrei Sura <andrei.sura@sharpspring.com>

## Description

When the number of rows in `schema_version` table is very large and keeps growing constantly due cronjobs dropping and re-creating tables we observed increase in memory usage.

It is not guaranteed this is the only issue but I thought this PR can help starting a discussion.

Here is what vttablet RAM looks like for ~1500 schema changes (we have more than 300 tables so every row has a large `schemax` column) :
Screen Shot 2021-08-16 at 11 07 09 AM
 

```
desc schema_version;
+--------------+------------------+------+-----+---------+----------------+
| Field        | Type             | Null | Key | Default | Extra          |
+--------------+------------------+------+-----+---------+----------------+
| id           | int(11)          | NO   | PRI | NULL    | auto_increment |
| pos          | varbinary(10000) | NO   |     | NULL    |                |
| time_updated | bigint(20)       | NO   |     | NULL    |                |
| ddl          | blob             | NO   |     | NULL    |                |
| schemax      | longblob         | NO   |     | NULL    |                |
+--------------+------------------+------+-----+---------+----------------+

-- query used to count the version rows

select count(*) from _vt.schema_version;
+----------+
| count(*) |
+----------+
|     2117 |
+----------+

select max(length(schemax)) from schema_version;
+----------------------+
| max(length(schemax)) |
+----------------------+
|               366580 |
+----------------------+

Total MB loaded in RAM
select 2117 * 366580/ (1024 * 1024);
+------------------------------+
| 2117 * 366580/ (1024 * 1024) |
+------------------------------+
|                     740.0988 |
+------------------------------+
```

I did some benchmarking using a mock of the code and measured the memory usage of the two blocks of code, and the numbers look promising to me
```
== old code

Memory usage in loadFromDB: {
allocated:1226 MiB 
totalAllocated:%!f(uint64=1857) MiB 
system:%!f(uint64=1358) MiB
heapAlloc:%!f(uint64=1226) MiB}

== new code	
Memory usage in loadFromDB: {
allocated:745 MiB
totalAllocated:%!f(uint64=1891) MiB
system:%!f(uint64=1153) MiB 
heapAlloc:%!f(uint64=745) MiB}
```

## Checklist
- [ ] Should this PR be backported?
- [ ] Tests were added or are not required
- [x] Documentation was added or is not required

### CODE REVIEW

File: go/vt/vttablet/tabletserver/schema/historian.go [hunk 1]
```diff
@@ -178,6 +182,13 @@ func (h *historian) loadFromDB(ctx context.Context) error {
 		h.lastID = id
 	}
 	h.sortSchemas()
+
+	loadedSchemas := len(h.schemas)
```
Comment (indera-shsp) [inline go/vt/vttablet/tabletserver/schema/historian.go:186]: loadFromDB is called every time schema changes and the number of entries in `h.schemas` can grow pretty fast

File: go/vt/vttablet/tabletserver/schema/historian.go [hunk 2]
```diff
@@ -178,6 +182,13 @@ func (h *historian) loadFromDB(ctx context.Context) error {
...
+	if loadedSchemas > maxVersions {
+		// keep only last N versions
+		h.schemas = h.schemas[loadedSchemas-maxVersions : loadedSchemas]
```
Comment (indera-shsp) [inline go/vt/vttablet/tabletserver/schema/historian.go:189]: @vmg If you have time to read this... can this cause a memory leak?

Comment (derekperkins) [reply go/vt/vttablet/tabletserver/schema/historian.go:189]: I'm pretty sure that isn't saving memory. It's still a slice on the same backing array and this is just moving the window.

Comment (indera-shsp) [reply go/vt/vttablet/tabletserver/schema/historian.go:189]: From the article linked the fix is to
```
 is to nil out the trailing pointer(s) before you truncate the slice.
```

Comment (indera-shsp) [reply go/vt/vttablet/tabletserver/schema/historian.go]: @derekperkins Is the new version acceptable?

**Comment by indera-shsp** [comment]
@deepthi Can you please recommend somebody to review this PR? 

Thank you!

**Comment by indera-shsp** [comment]
I wrote this to emulate the fix

```
package main

import (
	"fmt"
	"sync"
)

func main() {
	print("\n==> Old code")
	testHistorianOrig()
	print("\n==> New code")
	testHistorian()
}

type trackedSchema struct {
	//schema map[string]*binlogdatapb.MinimalTable
	//schema map[string]*string
	ddl    string
}

type historian struct {
	schemas []*trackedSchema
	mu      sync.Mutex
	enabled bool
	isOpen  bool
}

func newTrackedSchema(ddl string) *trackedSchema {
	r := &trackedSchema{
		ddl: ddl,
	}
	return r
}

func newHistorian(enabled bool) *historian {
	sh := historian{
		enabled: enabled,
	}
	return &sh
}


func (h *historian) loadFromDbOriginal(round int) error {
	loadVersions := 3

	for i := 1; i <= loadVersions; i++ {
		// strconv.Itoa(i)
		ddl := fmt.Sprintf("create tbl x_%d_%d", round, i)
		h.schemas = append(h.schemas, newTrackedSchema(ddl))
	}

	return nil
}

func (h *historian) loadFromDB(round int) error {
	loadVersions := 3

	for i := 1; i <= loadVersions; i++ {
		ddl := fmt.Sprintf("create tbl x_%d_%d", round, i)
		h.schemas = append(h.schemas, newTrackedSchema(ddl))
	}

	// fix
	maxVersions := loadVersions + 2
	loadedSchemas := len(h.schemas)
	if loadedSchemas > maxVersions {
		cutOff := loadedSchemas-maxVersions
		for i := 0; i < cutOff; i++ {
			h.schemas[i] = nil
		}
		h.schemas = h.schemas[cutOff : loadedSchemas]
	}

	return nil
}

func (h *historian) printMe() {
	loadedSchemas := len(h.schemas)

	str := fmt.Sprintf("\nHistorian - loadedSchemas: %d", loadedSchemas)
	str += fmt.Sprintf("\n\t %s", h.schemas)
	print(str)
}

func testHistorianOrig() {
	h := newHistorian(true)
	h.loadFromDbOriginal(1)
	h.printMe()

	h.loadFromDbOriginal(2)
	h.printMe()

	h.loadFromDbOriginal(3)
	h.printMe()
}

func testHistorian() {
	h := newHistorian(true)
	h.loadFromDB(1)
	h.printMe()

	h.loadFromDB(2)
	h.printMe()

	h.loadFromDB(3)
	h.printMe()
}

```

**Comment by sonne5** [comment]
Having `maxVersions` requires ppl know how many schema migrations they want for all the binlogs, otherwise we might remove the an old schema so is required for vstream's schema tracking. 

We are having similar discussions, and I think a more ergonomic fix should be just load schema version that is enough to cover all the binlogs https://github.com/vitessio/vitess/issues/8585

**Comment by indera-shsp** [comment]
> Having `maxVersions` requires ppl know how many schema migrations they want for all the binlogs, otherwise we might remove the an old schema so is required for vstream's schema tracking.
> 
> We are having similar discussions, and I think a more ergonomic fix should be just load schema version that is enough to cover all the binlogs #8585

@sonne5 
Does your comment assume that covering all the binlogs _always_ leads to a small and safe number of rows loaded to RAM?

**Comment by sonne5** [comment]
> > Having `maxVersions` requires ppl know how many schema migrations they want for all the binlogs, otherwise we might remove the an old schema so is required for vstream's schema tracking.
> > We are having similar discussions, and I think a more ergonomic fix should be just load schema version that is enough to cover all the binlogs #8585
> 
> @sonne5
> Does your comment assume that covering all the binlogs _always_ leads to a small and safe number of rows loaded to RAM?

Whether if is small depends on how frequent the schema change is, but it should be safe (as in not break the existing functionalities)

**Comment by indera-shsp** [comment]
> > > Having `maxVersions` requires ppl know how many schema migrations they want for all the binlogs, otherwise we might remove the an old schema so is required for vstream's schema tracking.
> > > We are having similar discussions, and I think a more ergonomic fix should be just load schema version that is enough to cover all the binlogs #8585
> > 
> > 
> > @sonne5
> > Does your comment assume that covering all the binlogs _always_ leads to a small and safe number of rows loaded to RAM?
> 
> I think whether if is small depends on how frequent the schema change is, but it should be safe (as in not break the existing functionalities)

It sounds like you are advocating for this to be configurable?

**Review by deepthi — COMMENTED** [review]
Apologies that reviewing this took so long.
The code itself looks fine. You seem to have done a lot of due diligence on this PR, which is definitely appreciated.

I believe to address the concerns brought up so far:
- maxVersions should be configurable via a flag and should default to the current number of 10000
- a test case is required to ensure the flag is being respected. you can simply convert your standalone test into a unit test.

Auto-tuning this number to cover the binlog period as described in #8585 can be a future improvement.

**Pull request closed without merging**